src/statistics/MetricsCompressor.ts

```diff
@@ -0,0 +1,382 @@
+
+/**
+ * This class generates binary representation of client metrics
+ * (i.e. numeric statistics).
+ * @internal
+ */
+export class MetricsCompressor {
+
+    private readonly metricsBuffer: OutputBuffer;
+    private readonly dictionaryBuffer: OutputBuffer;
+    private readonly dictionary: MetricsDictionary;
+    private metricsCount = 0;
+    private lastDescriptor: MetricDescriptor;
+
+    constructor() {
+        this.metricsBuffer = new OutputBuffer();
+        this.dictionaryBuffer = new OutputBuffer();
+        this.dictionary = new MetricsDictionary();
+    }
+
+    addLong(descriptor: MetricDescriptor, value: number): void {
+        this.writeDescriptor(descriptor);
+        this.metricsBuffer.writeByte(ValueType.LONG);
+        this.metricsBuffer.writeLong(value);
+    }
+
+    addDouble(descriptor: MetricDescriptor, value: number): void {
+        this.writeDescriptor(descriptor);
+        this.metricsBuffer.writeByte(ValueType.DOUBLE);
+        this.metricsBuffer.writeDouble(value);
+    }
+
+    generateBlob(): Promise<Buffer> {
+        this.writeDictionary();
+
+        const metricsBuf = this.metricsBuffer.toBuffer();
+        const dictionaryBuf = this.dictionaryBuffer.toBuffer();
+
+        return Promise.all([
+            this.compressBuffer(metricsBuf),
+            this.compressBuffer(dictionaryBuf)
+        ]).then(([compressedMetricsBuf, compressedDictionaryBuf]) => {
+            const completeSize = SIZE_VERSION
+                + SIZE_DICTIONARY_BLOB + compressedDictionaryBuf.length
+                + SIZE_COUNT_METRICS + compressedMetricsBuf.length;
+            const finalBuf = new OutputBuffer(completeSize);
+            finalBuf.writeByte((BINARY_FORMAT_VERSION >>> BITS_IN_BYTE) & BYTE_MASK);
+            finalBuf.writeByte(BINARY_FORMAT_VERSION & BYTE_MASK);
+            finalBuf.writeInt(compressedDictionaryBuf.length);
+            finalBuf.writeBuffer(compressedDictionaryBuf);
+            finalBuf.writeInt(this.metricsCount);
+            finalBuf.writeBuffer(compressedMetricsBuf);
+            return finalBuf.toBuffer();
+        });
+    }
+
+    private writeDescriptor(descriptor: MetricDescriptor): void {
+        const mask = this.calculateDescriptorMask(descriptor);
+        this.metricsBuffer.writeByte(mask);
+
+        if ((mask & MASK_PREFIX) === 0) {
+            this.metricsBuffer.writeInt(this.getDictionaryId(descriptor.prefix));
+        }
+        if ((mask & MASK_METRIC) === 0) {
+            this.metricsBuffer.writeInt(this.getDictionaryId(descriptor.metric));
+        }
+        if ((mask & MASK_DISCRIMINATOR) === 0) {
+            this.metricsBuffer.writeInt(this.getDictionaryId(descriptor.discriminator));
+        }
+        if ((mask & MASK_DISCRIMINATOR_VALUE) === 0) {
+            this.metricsBuffer.writeInt(this.getDictionaryId(descriptor.discriminatorValue));
+        }
+        if ((mask & MASK_UNIT) == 0) {
+            this.metricsBuffer.writeByte(descriptor.unit);
```

Comment:
Good finding. Initially I assumed the `unit` is always present, but found that it may be `null` in Java client's stats gauges. Forgot to update this part of the code accordingly. Fixed in dd2a2e094520d0c039e064720bc61ee5de8ec71c